AST transforms for comprehensions

wcjohnson opened the pull request:
This PR includes a complete version of the implicit-returns PR. I will rebase after that one gets merged.

wcjohnson commented:
@rattrayalex 

Big fixes. Closures work now!

rattrayalex commented on src/index.js:
```diff
@@ -183,20 +183,56 @@ export default function (babel) {
+      if (tailPath.isExpressionStatement()) {
+        visitor(tailPath);
+      } else if (tailPath.isFunctionDeclaration()) {
+        // Promote function declarations to exprs.
```
Bah, so, here's an edge case that this might break: 

```
outer(x) ->
  now x = inner(x)

  inner() ->
    x
```

obviously, this is silly code, but a function that returns a function – and uses that function internally – needs the returned function to be declared, so that it is hoisted. Converting it to an expression breaks that. 

Okay to stick with the `return fnName` semantic from the other PR?

rattrayalex replied:
If so, that should obviate this function I think...

wcjohnson replied:
Are we sure this is broken? Look at the output code. Remember that ultimately it's the javascript engine that runs it, so as long as the output code is correct, it will work.

I think a `FunctionExpression` with a name gets pretty printed by Babel into a declaration in the final code. That means that a modern JS engine will let you call it at a higher point in the scope because that's what ES5 says. So I actually think the output is correct.

One way or another, we definitely can't just blindly insert `return` as we do with internal functions -- because remember, these exprs are going to be `replacedWith` things like `array.push` or `obj[k]=v`. I'm not sure how those two things would square with each other. We would probably have to wrap in an IFE.

My inclination is that if this is generating correct code at the engine level we shouldn't change it.

wcjohnson replied:
The core reason this code exists is because you can't insert anything that's not an expr into the `CallExpresion` for array.push or the `AssignExpression` for obj[k]=v without failing Babel validation. So whatever we return from here it has to be an expression, and there isn't anyway around that that I can see.

rattrayalex replied:
Try these in your console: 

```
function outer(x) {
  x = inner();
  function inner() {
    return x + 1
  }
  return inner
}
outer(1)()
// 3
```
and
```
function outer2(x) {
  x = inner();
  return function inner() {
    return x + 1
  }
}
outer2(1)()
// Uncaught ReferenceError: inner is not defined
```

rattrayalex replied:
we should be generating the former.

rattrayalex replied:
> I think a FunctionExpression with a name gets pretty printed by Babel into a declaration in the final code.

I don't believe so; Babel will either wrap it in parentheses if it's not already part of some expression.

rattrayalex replied:
For the expression case, we can do the same thing that we're doing in `addImplicitReturns`, which is to put the name of the function in the arr.push / obj[k] =

so for example: 
```
{ for elem x in arr: x, f() -> x }
```
becomes
```
const _obj = {};
for (...) {
  const x = arr[_i];
  function f() { return x }
  _obj[x] = f;
}
```
thoughts?

wcjohnson replied:
Right, this is exactly why that code is needed!

Check out the output of the two test fixtures I just pushed. AFAICT his PR is generating the proper final js output for interior declared functions as of right now.

wcjohnson replied:
@rattrayalex Babel's output plugin already does this automatically I think... Look at the new fixtures, they correctly move the declaration up

rattrayalex replied:
Right, sorry, the comparable test case for comprehensions would be: 

```
[for elem x in arr:
  y = f(x)
  f(z) -> y + z
]
```
is that right?

rattrayalex replied:
maybe I'm missing something... could help if you can show a test case that fails with the FunctionExpression transform removed...

wcjohnson replied:
@rattrayalex Nevermind, you're right, the double nested function was making it work. It doesn't work at one level of nesting.

rattrayalex replied:
Cool! That makes sense

rattrayalex commented on src/index.js:
```diff
@@ -183,20 +183,56 @@ export default function (babel) {
-    let completionRecords = path.get("body").getCompletionRecords();
-    for (let targetPath of completionRecords) {
-      if (!targetPath.isExpressionStatement()) continue;
+  function validateComprehensionLoopBody(loopBodyPath) {
```
nice, I like this approach

rattrayalex commented on src/index.js:
```diff
@@ -183,20 +183,56 @@ export default function (babel) {
-      );
-      targetPath.replaceWith(arrPush);
-    }
+  function wrapComprehensionInIFE(bodyVarId, bodyVarInitializer, loopBody) {
```
nice!

rattrayalex replied:
maybe `wrapComprehensionInIife` ?

wcjohnson replied:
Sure.

rattrayalex commented on src/index.js:
```diff
@@ -632,6 +668,16 @@ export default function (babel) {
 
+  definePluginType("ObjectComprehension", {
+    visitor: ["loop"],
+    aliases: ["Expression"],
```
Should probably also alias "ObjectExpression", no?

wcjohnson replied:
Yes, will change.

rattrayalex commented on src/index.js:
```diff
@@ -883,34 +929,49 @@ export default function (babel) {
-          ReturnStatement(returnPath) {
-            throw returnPath.buildCodeFrameError("`return` is not allowed within Comprehensions.");
-          },
+        validateComprehensionLoopBody(path.get("loop.body"));
```
I was going to say the validation could go into an `enter` while the tail expression processing goes into an `exit`, but I actually think the validation would be better off in an `exit` too. That way it'll validate `<-` properly in addition to `await` etc – thoughts?

wcjohnson replied:
Not sure this will work. Regardless of enter or exit, the visitor is being called at a point-in-time -- exactly when we call `validateComprehensionLoopBody`, so if the processing hasn't been done by then it won't matter.

A possible change here is to move the whole comprehension visitor to exit.

rattrayalex replied:
yeah, moving the comprehension visitors to exit I think would be the way to go

wcjohnson replied:
Moving to exit visitor caused some pretty horrific regressions. Stuff like this:
```
Error: /Users/wcj/dev/node_modules/babel-plugin-lightscript/test/fixtures/comprehensions/closure-nested-semantic.js: We don't know what to do with this node type. We were previously a Statement but we can't fit in here?
      at NodePath.insertBefore (node_modules/babel-traverse/lib/path/modification.js:61:13)
```

wcjohnson replied:
@rattrayalex 

Yup, I think this can't be an exit visitor. (At least not without further extensive modifications.) The `ForInArrayStaetement` gets transformed first and that breaks the assumptions we're making about ast pretty badly.

I have a working implemenation as an enter visitor which I think I will stick with for this PR.

rattrayalex replied:
Ah, okay, that makes sense... too bad

rattrayalex commented on src/index.js:
```diff
@@ -883,34 +929,49 @@ export default function (babel) {
+        validateComprehensionLoopBody(path.get("loop.body"));
+
+        const id = path.scope.generateUidIdentifier("arr");
+        visitTailExpressions(path.get("loop"), true, function(tailPath) {
```
As noted elsewhere, I don't think `visitTailExpressions` is needed anymore – a plain `for-of` here would probably be simpler

wcjohnson replied:
I wouldn't be too sure about this, see my comments above.

rattrayalex commented on src/index.js:
```diff
@@ -883,34 +929,49 @@ export default function (babel) {
         });
 
-        const arrId = path.scope.generateUidIdentifier("arr");
+        path.replaceWith(wrapComprehensionInIFE(id, t.arrayExpression(), path.node.loop));
```
nice, this should make it easier to "unroll" in the future

rattrayalex commented on src/index.js:
```diff
@@ -883,34 +929,49 @@ export default function (babel) {
+            (seqExpr.expressions.length !== 2)
+          ) {
+            throw tailPath.buildCodeFrameError("Object comprehensions must return" +
+            " sequence expressions of the form `key`, `value`.");
```
How about
`Object comprehensions must return a (key, value) pair.`

wcjohnson replied:
Been thinking about this a bit, don't think we should say `return` because then people might be tempted to actually write `return`. Also don't think we can say "in tail position" because most people don't know what that means. Do agree that it should change a bit but not sure on final wording.

rattrayalex replied:
Hmm, good points, how about... 
`Object comprehensions must end with a (key, value) pair.`

wcjohnson replied:
yeah that is probably the best possible, will change.

rattrayalex reviewed (commented):
Looking great – I think basically one real change

wcjohnson commented:
@rattrayalex 

Addressed some comments, added new fixtures related to the interior declared functions. I believe the current code generation is both syntactically and semantically correct.

rattrayalex commented:
@wcjohnson once those fixes go in, I think this will be ready to go – you might want to squash! hoping to merge shortly 😄

wcjohnson committed:
AST transforms for comprehensions

Squashed commits:
[3d0fdee] isExpression
[29394f0] Fixtures for assignments and declarations in comprehensions
[4bb8f47] AST transforms for comprehensions

Squashed commits:
[34ed4e7] Fix for object comprehensions and inner function decls + associated fixtures
[265fa46] Object comprehensions not working with inner declarations…
[0f484d8] AST transforms for comprehensions.

(Squashed 23 commits)

wcjohnson commented:
@rattrayalex 

Fixes are in.

wcjohnson commented:
Strike that, still an issue.

wcjohnson commented:
This is a nontrivial problem for object comprehensions. The tail expression is a `SequenceExpression`, but if the second expression in the sequence is a `FunctionExpression.` It needs to be turned back into a Declaration and hoisted using the old approach I was using. Won't be done tonight.

wcjohnson commented:
Or maybe it will?

rattrayalex commented:
I think if it starts as a `FunctionExpression` it should stay as one. The edge case I outlined for implicit returns / array comprehensions isn't one that I mean to encourage; just something that shouldn't break due to unnecessary AST mangling. 

In general, let's try to keep modification to a minimum.

wcjohnson commented:
@rattrayalex 

OK, so to be clear, if someone writes:
```
{for idx i in Array(3):
  g()
  (i, g() -> i)
}
```
and then get a runtime error about g being undefined, that is the desired behavior?

Whereas if they do the same thing with an array comprehension it will work?

wcjohnson commented:
For the record, my own opinion is I'm fine with neither of those things working, or both of them working, but one working and another not for no apparent reason is something that bugs me a bit.

Mangling the AST is a foregone conclusion if we decide to support this pattern. We are already `insertBefore` or `insertAfter` a declaration or return in the implicit return case and the array case. This is essentially the exact same mangling.

rattrayalex commented:
Hmm, this is tricky. Let me think about it. 
(you mean `insertAfter` I think btw)

rattrayalex commented:
So, hoisted function declarations are a feature of JS that you probably have to know about to use. 

And people who use them are very likely to know that you can only hoist a top-level function (they may not think about it as a declaration/statement vs expression, but you know it can't be within something else). 

The form `a, b` is a little weird – it's like a faux-tuple – but it definitely looks like "a thing". And you can't declare a hoisted function "within a thing". 

I don't think people would be surprised to see this not work: 
```
  foo()
  (x, function foo() {...})
```
but I do think they would be surprised to see this not work:
```
  foo()
  function foo() {...}
```

the former says, "I'm sticking a function inside of this thing", the latter says, "I'm just declaring a function!"

LightScript generally seeks to sit *on top* of JavaScript – not to mess around with it. There are obviously times when that's not true, but it's kept to a minimum. 

Keeping Functions as declarations/expressions is following that spirit. 

Does that make sense?

rattrayalex commented:
tl;dr, 

>  for no apparent reason

I think the reason is apparent

rattrayalex commented on src/index.js:
```diff
@@ -883,34 +942,56 @@ export default function (babel) {
+        validateComprehensionLoopBody(path.get("loop.body"));
+
+        const id = path.scope.generateUidIdentifier("arr");
+        transformTails(path.get("loop"), true, function(expr) {
```
would prefer an arrow function but nbd

rattrayalex commented:
Other than the FunctionDeclaration/Expression bit, this looks good to go!

wcjohnson commented:
Again for the record, I disagree philosophically, because we are letting people violate the JS pattern elsewhere. For instance, we know in regular JS:
```javascript
function f() {
  g()
  return function g() { }
}
```
doesn't work, because `function g()` is in expr context. Makes sense, fine and dandy.

But in LightScript, we are now allowing:
```javascript
f() ->
  g()
  g() -> {}
```
and it works fine. If we're thinking "light wrapper over JavaScript" most people would expect that to fail for the exact same reason, no? From the point of view of the actual user of the language, we are already turning expressions into declarations, whether or not the internals are actually doing that or not.

However, I think we are talking about the edgiest of edge cases here. We are basically talking about code that people should probably be shot for writing -- so I will revert that change.

rattrayalex commented:
Cool, thanks. 

Yeah, this is sort of a problem with the whole idea of implicit returns... sometimes you want them, sometimes you don't. I'm really hoping to build tooling that encourages `: void` annotations whenever the retval isn't used. 

The problem with the implicit returns case is: 

```coffee
f() ->
  g()
  g() -> {}
```
the user might not have meant to return `g`; they might not be thinking of it as 
```js
function f() {
  g()
  return function g() { }
}
```
at all, but instead just as
```js
function f() {
  g()
  function g() { }
}
```

Of course, that breaks down when it comes to comprehensions, where the ending-thing is clearly going to be "returned". 

Anyway, you're definitely right that this isn't an edge-case worth spending much time on; which is largely why I'd really like the simplest/smallest/least intrusive solution. We need the "if its a declaration, return its name afterwards" code for implicit returns anyway, so reusing it when it's actually needed in comprehensions feels cleanest. 

Thanks for taking this on 😄 and excited to ship object comprehensions!

rattrayalex commented on src/index.js:
```diff
@@ -183,22 +183,70 @@ export default function (babel) {
+      } else if (isFunctionDeclaration(tailPath.node)) {
+        // Need to transform exprs to statements since this is block context.
+        const nextNode = getNewNode(tailPath.node.id, tailPath);
+        if (t.is("Expression", nextNode)) {
```
I think `t.isExpression` could be used next time, FYI (unless I'm missing something?) but I have no problem using this for now

wcjohnson replied:
@rattrayalex Not sure if you're missing anything, TBH. I was looking into the Babel code to introspect node types and I found `t.is`. I pulled on the thread and I thought if I called that method it would correctly understand the LightScript type aliases. (So I wouldn't need e.g.
```
    return node && (t.is("FunctionDeclaration", node) || node.type === "NamedArrowDeclaration");
```
to  find out if something is a function declaration.

However, it turns out it doesn't work, and I can't figure out why it doesn't work NamedArrowDeclaration definitely has FunctionDeclaration as an alias, and the code for `t.is` definitely checks the aliases. I was totally mystified.

I bet t.isExpression() would be fine here so I will amend.

rattrayalex replied:
Hmm, wouldn't this be necessary for the VariableDeclaration codepath too?

I guess the test case for that would be 
```
[for idx i in Array(10): 
  const x = f(i);
]
```
which reminds me we should definitely have a test for 
```
[for idx i in Array(10): 
  x = f(i);
]
```
which I have a hunch may not do what we want it to. 

Though I'm not sure in either case what we really want here. 

With implicit returns, it can be kind of nice to annotate your return value with a name – or at least, it can be annoying when doing so results in implicit returns turning off or breaking. I guess the same logic applies for comprehensions, so I think we want the same behavior, but it's a little weird (and should almost definitely be a linting error). 

If you just wanna get this shipped for now though I'm happy to merge and add the above in a follow-on

rattrayalex replied:
Also, apologies for not noticing this earlier 😕

wcjohnson replied:
My guess is the assignment expression will work but the declaration won't, will test.

wcjohnson replied:
They both worked without change:
```
[for idx i in Array(10):
  now x = f(i)
]
```
```javascript
(() => {
  const _arr = [];

  const _arr2 = Array(10);

  for (let i = 0, _len = _arr2.length; i < _len; i++) {
    x = f(i);

    _arr.push(x)
  }

  return _arr;
})();
```
and
```
[for idx i in Array(10):
  const x = f(i)
]
```
```javascript
(() => {
  const _arr = [];

  const _arr2 = Array(10);

  for (let i = 0, _len = _arr2.length; i < _len; i++) {
    const x = f(i);

    _arr.push(x)
  }

  return _arr;
})();
```

wcjohnson replied:
Looks right to me

rattrayalex commented on test/fixtures/comprehensions/assignment-expr/actual.js:
```diff
@@ -0,0 +1,3 @@
+[for idx i in Array(10):
+  now x = f(i)
```
Actually, with `now`, it's perfectly fine to leave it as `arr.push(x = f(i));` – the case I was worried about was the auto-const `x = f(i)`

rattrayalex replied:
I'm fine with this output too though if the code is simpler

wcjohnson replied:
I expect we can change the output of `now` to look like that very easily, but for some reason `addImplicitReturns` extracts the LHS of the assignment operator when returning it.

AFAIK it's perfectly fine to `return x=3` or some such, but if you had some reason for returning the LHS in `addImplicitReturns` it would be good to know why that's there, and/or why this should be the same and/or different.

Right now it's just using the same code path.

wcjohnson replied:
Auto const works fine afaict.

rattrayalex replied:
Cool. Yeah it's there in `addImplicitReturns` from before that was moved to an `exit()` visitor, when `x = 3` was going to turn into `const x = 3`. It actually shouldn't be needed for returns anymore, but now we're relying on it for this. Fine to leave as-is I think.

rattrayalex reviewed (approved):
Awesome! Looks great! Good by you if I merge?

wcjohnson commented:
@rattrayalex Squashed, good to merge as f ar as I can see.
